agenta-backend/agenta_backend/middlewares/api_key_manager.py

```diff
@@ -0,0 +1,229 @@
+        # Update the API key
+        engine.save(api_key)
+
+    def delete_api_key(self, key):
+        """
+        Deletes an API key with the provided key.
+
+        Args:
+        - key: The key of the API key to be deleted.
+
+        Raises:
+        - KeyError: If the API key does not exist.
+        """
+        # Check if the API key exists
+        existing_key = engine.find_one(APIKeyDB, APIKeyDB.key == key)
+        if not existing_key:
+            raise KeyError("API key not found")
+
+        # Delete the API key
+        engine.delete(existing_key)
+
+    def is_valid_api_key(self, key):
+        """
+        Checks if an API key is valid by verifying that it is not blacklisted and not expired.
+
+        Args:
+        - key: The API key to be checked.
+
+        Returns:
+        - True if the API key is valid, False otherwise.
+        """
+        # Check if the API key is valid (not blacklisted and not expired)
+        api_key = engine.find_one(APIKeyDB, APIKeyDB.key == key)
+        if not api_key:
+            return False
+
+        if api_key.blacklist:
+            return False
+
+        if api_key.expiration_date and api_key.expiration_date < datetime.utcnow():
+            return False
+
+        return api_key
+
+    async def use_api_key(self, key):
+        """
+        Uses an API key and checks rate limiting.
+
+        Args:
+        - key: The API key to be used.
+
+        Returns:
+        - True if the API key can be used, False otherwise.
+        """
+        # Use the API key and check rate limiting
+        api_key = self.is_valid_api_key(key)
+        if not api_key:
+            raise ValueError("Invalid API key, make sure it exists, is not blacklisted or expired")
+
+        try:
+            if api_key.rate_limit > 0:
+
+                # Check rate limiting in Redis
+                if self.redis_client is None:
+                    await self.init_redis()
+
+                cache_key = f"api_request_count:{key}"
+                api_requests_within_minute = await self.redis_client.get(cache_key)
+
+                if api_requests_within_minute is None:
+                    # Initialize the count in Redis with an initial value of 1 and a one-minute TTL
+                    await self.redis_client.setex(cache_key, 60, 1)
+                else:
+
+                    # Check if requests made within the last minute exceed the rate limit
+                    count_within_minute = int(api_requests_within_minute)
+                    if count_within_minute > api_key.rate_limit:
+                        raise ValueError("API key rate limit exceeded")
+
+            # Update the last usage timestamp for rate limiting
+            api_key.updated_at = datetime.utcnow()
+            engine.save(api_key)
+
+            return True
+        except Exception as e:
+            logger.error(f"Error using API key: {e}")
+            raise e
+
+
+async def authenticate_middleware(request: Request, call_next):
```

Comment:
The logic here sounds reasonable. Please test this before raising the next PR (in ee repo). To test it create first a simple function **outside** of agenta-backend. Use the middleware and see if it works by itself, then you can add it to agenta for one endpoint (while removing supertokens) and see if it works, last please try to integrate it with supertokens